### PR TITLE
Feature/new binding opcua

### PR DESCRIPTION
**Pull request by erossignon**

### CODE REVIEW

**Comment by erossignon** [comment]
> Some quick notes, not a review.
> 
> * This PR introduces packages/binding-opcua2 . That is not what we want in the end. There should be just _one_ opcua
> * It changes parts that do not belong to this or
>   
>   * http
>   * coap
>   * mbus
>   * websocket
>   * core
>   * examples
>   * td-tools

This mainly is related to updating the "wot-typescript-definition" package that was needed for the new binding to fix a typescript issue.  To be consistent, the "wot-typescript-definition" needed to be the same for all submodule that uses it . 
I have now extracted this  "wot-typescript-definition"  version bump to a different PR.

**Comment by erossignon** [comment]
> * This PR introduces packages/binding-opcua2 . That is not what we want in the end. There should be just _one_ opcua

Sure, we will do this once Christiano and Lucas have validated that the binding-opcua2 can replace the current binding-opcua in their app.

**Comment by erossignon** [comment]
I have now separated code changes not related to binding-opcua2 into separate PR.
Those ones (#663 #667 )  need to be merged before we can merge this one.
 #665 #666 are addition  fixes.

**Comment by erossignon** [comment]
The PR is ready for review and integration in main branch

**Comment by danielpeintner** [comment]
Note: the test pipeline issue for "setup-and-test (windows-latest, 12.x)" seems unrelated.

File: packages/binding-opcua/README.md
Comment (danielpeintner) [inline packages/binding-opcua/README.md:66]: Do these files actually exist?

File: packages/binding-opcua/README.md
Comment (danielpeintner) [inline packages/binding-opcua/README.md:105]: I think this is this line is not correct.

File: packages/binding-opcua/src/opcua_protocol_client.ts
Comment (danielpeintner) [inline packages/binding-opcua/src/opcua_protocol_client.ts:1]: naming conventions.

you seem to use it inconsistently  "_"  and "- "in the opcua binding package.
We usually use hyphens.
* opcua_protocol_client.ts --> opcua-protocol-client.ts
* client_test.ts --> client-test.ts
* ...

**Comment by danielpeintner** [comment]
@lukesmolo @relu91 @wiresio may I ask you to take a look at the PR..

File: packages/binding-opcua/src/opcua_protocol_client.ts
Comment (erossignon) [reply packages/binding-opcua/src/opcua_protocol_client.ts:1]: Sorry, I am using the opposite convention in node-opcua. I need to play a cameleon and use kebab-case here instead of _snake_case.

File: packages/binding-opcua/src/codec.ts
Comment (relu91) [inline packages/binding-opcua/src/codec.ts:164]:
```suggestion
        const wantDataValue = (parameters?.to === "DataValue");
```

File: packages/binding-opcua/src/opcua-protocol-client.ts
Comment (relu91) [inline packages/binding-opcua/src/opcua-protocol-client.ts:53]: Do we need to export this type?

File: packages/binding-opcua/src/opcua-protocol-client.ts
Comment (relu91) [inline packages/binding-opcua/src/opcua-protocol-client.ts:88]: also here, do we need export?

File: packages/binding-opcua/src/opcua-protocol-client.ts
Comment (relu91) [inline packages/binding-opcua/src/opcua-protocol-client.ts:293]: throwing should be fine 👍🏻

File: packages/binding-opcua/src/opcua-protocol-client.ts
Comment (relu91) [inline packages/binding-opcua/src/opcua-protocol-client.ts:453]: `_contentToDataValue2` why the two?

File: packages/core/src/content-serdes.ts
Comment (relu91) [inline packages/core/src/content-serdes.ts:33]: Not sure this was meant to be exposed to bindings, why did you need it?

**Review by relu91 — CHANGES_REQUESTED** [review]
I've just commented what I can. I didn't review in detail the opc-ua client logic, I leave that to @lukesmolo (which is now pretty busy but he promised to look into this asap). There are minor changes required and then it looks ok for me.

File: packages/binding-opcua/package.json
Comment (erossignon) [reply packages/binding-opcua/package.json:41]: ajv-format should be in "dependencies" as well

File: packages/binding-opcua/package.json
Comment (erossignon) [reply packages/binding-opcua/package.json:57]: DataSchemaValue (exposed by wot-typescript-definitions) is used in the codec file.
so wot-typescript-definitions should be in "dependency"

File: packages/binding-opcua/src/opcua-protocol-client.ts
Comment (erossignon) [reply packages/binding-opcua/src/opcua-protocol-client.ts:88]: no, indeed, thanks

File: packages/binding-opcua/src/opcua-protocol-client.ts
Comment (erossignon) [reply packages/binding-opcua/src/opcua-protocol-client.ts:293]: OK

File: packages/binding-opcua/src/opcua-protocol-client.ts
Comment (erossignon) [reply packages/binding-opcua/src/opcua-protocol-client.ts:453]: Let me simplify this.

**Comment by wiresio** [comment]
@erossignon - looks like you removed the Readme. Could you please tell me how to run the tests?

`ts-node opcua-server.ts` doesn't do anything.

`ts-node opcua-client-test.ts` results in:

```
/home/developer/git/thingweb.node-wot/packages/binding-opcua/test/opcua-client-test.ts:35
describe("OPCUA client test", function () {
^
ReferenceError: describe is not defined
    at Object.<anonymous> (/home/developer/git/thingweb.node-wot/packages/binding-opcua/test/opcua-client-test.ts:35:1)
    at Module._compile (internal/modules/cjs/loader.js:1085:14)
    at Module.m._compile (/usr/lib/node_modules/ts-node/src/index.ts:1459:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Object.require.extensions.<computed> [as .ts] (/usr/lib/node_modules/ts-node/src/index.ts:1462:12)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:12)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:75:12)
    at main (/usr/lib/node_modules/ts-node/src/bin.ts:389:12)
    at Object.<anonymous> (/usr/lib/node_modules/ts-node/src/bin.ts:539:3)

```

**Comment by erossignon** [comment]
- binding-opcua/test/xxx_test.ts files are mocha tests that can be run using the  ` npm run test ` command

- standalone examples are in `examples/src/opcua`  folder

- The readme file has been update in this PR: see https://github.com/node-opcua/thingweb.node-wot/tree/Feature/NewBindingOPCUA/packages/binding-opcua

File: packages/binding-opcua/README.md
Comment (danielpeintner) [inline packages/binding-opcua/README.md:167]: Minor: I think this entire line is not needed. The link is already provided in the paragraph before that.

**Comment by wiresio** [comment]
_standalone examples are in examples/src/opcua folder_ - don't have that folder. Is it in another branch?

File: packages/binding-opcua/README.md
Comment (erossignon) [reply packages/binding-opcua/README.md]: OK fixed.

File: packages/binding-opcua/package.json
Comment (relu91) [reply packages/binding-opcua/package.json:57]: No, if you just using it as a type it should not be under "dependency". `wot-typescript-definitions` are just types (like those from strictly typed repository: `@type/node` or `@type/node-fetch`) and usually types are under `devDependencies`.

File: packages/binding-opcua/test/full-opcua-thing-test.ts
Comment (relu91) [inline packages/binding-opcua/test/full-opcua-thing-test.ts:392]: I was reviewing again the PR and this detail got me worried. Why are you calling `valueOf`?? the default behavior should be that the value function returns an application value not opc-ua payload information. @danielpeintner what do you think?

**Review by relu91 — CHANGES_REQUESTED** [review]

File: packages/binding-opcua/test/full-opcua-thing-test.ts
Comment (danielpeintner) [reply packages/binding-opcua/test/full-opcua-thing-test.ts:392]: I tend to agree. I mean for primitive values like strings it does not matter but for an object it makes a a difference. 
Is there a reason for calling `valueOf()` we are not aware of? (for the test?)

Comment (erossignon) [reply packages/binding-opcua/test/full-opcua-thing-test.ts]: valueOf() doesn't seem to be needed at all, I've cleaned this.

**Review by relu91 — APPROVED** [review]
Looks fine, even if I didn't check every single detail (due to the amount of line changed) I think we can move on and fix it later.

**Review by danielpeintner — APPROVED** [review]
Looks good to me.

Since I am not familiar with OPCUA.
Shall we wait for reviews from
@lukesmolo and @wiresio ?

**Comment by relu91** [comment]
I talked with @lukesmolo and I would be happy to try it if we can re-introduce the example default-servient that we used to have in the example folder. I would speed up his tests. He tried to create it but he stumbled upon some issue. So he asked to me if it would be possible to introduce it.. @danielpeintner we decided to remove this kind of default servients but maybe we can have an exception just for the sake of this PR. What do you think?

**Comment by danielpeintner** [comment]
> @danielpeintner we decided to remove this kind of default servients but maybe we can have an exception just for the sake of this PR. What do you think?

Fine by me.
Do you want to do this within this PR? Who shall do it? @lukesmolo  or @erossignon ?

BTW, @erossignon your last commit was not properly signed. Hence the Eclipse  ECA check fails.

**Comment by relu91** [comment]
I would ask @erossignon to add it in this PR if is not too much a trouble.

**Comment by wiresio** [comment]
@danielpeintner - let's merge it and release node-wot v0.8

**Pull request merged**
